Port zenoh_c to zenoh_cpp

ahcorde opened the pull request:
For visibility reason I created a branch called `dev/1.0.0` in this repository, otherwise I can't open the PR 

The goal of this PR is to port the `zenoh_c` API to `zenoh_cpp`.

I added a hack in the zenoh_cpp API to be able to make the transition easier. Edit the following file:

```bash
nano +83 install/opt/zenoh_cpp_vendor/include/zenoh/api/base.hxx
```

And add this change:

```cpp
public:
    OwnedType _0;
protected:
```
------

 - zenohd is ported to `zenoh_cpp`

ahcorde committed:
zenoh_c to zenoh_cpp

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Fixed CI

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

tokens

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

removed zenoh_router_check

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

removed rmw_init_options_impl.hpp

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Publisher

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Added subscribers

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Removed zid_to_str

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

use zenoh::KeyExpr

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

service data

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

client data

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Removed unused signature

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

cleanups and shm

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

more changes

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Fixed segfault

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

make linters happy

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

removed zenohc from CMakeLists.txt and warning

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Restored one method

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

wrap the zenoh session with a shared pointer

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

Changes missed

Signed-off-by: Yadunund <yadunund@gmail.com>

ahcorde commented:
| Pacakge                 | Test                                                                  | Reason                                                    |
| -                       | -                                                                     | -                                                         |
| rcl | None                             |                                            |
| rcl_action | None                             |                                            |
| rcl_lifecycle | None                             |                                            |
| rcl_yaml | None                             |                                            |
| rcl_yaml_param_parser | None                             |                                            |
| rclcpp | test_executors                             |                                            |
| rclcpp_components | test_executors                             |                                            |
| rclcpp_action | None                             |                                            |
| rclcpp_lifecycle | None                             |                                            |
| rclcpp_components | None                             |                                            |

# rclcpp
## test_executors
```
95: /home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/executors/test_executors.cpp:759: Failure
95: Expected equality of these values:
95:   sub2_msg_count
95:     Which is: 1
95:   1u
95:     Which is: 1
````

ahcorde commented:
> @ahcorde would it be possible to rebase this branch such that it only includes the newly introduced commits?

@Yadunund It's already done

Yadunund commented on zenoh_cpp_vendor/CMakeLists.txt:
```diff
@@ -35,4 +35,15 @@ ament_vendor(zenoh_c_vendor
 
 ament_export_dependencies(zenohc)
 
+ament_vendor(zenoh_cpp_vendor
```
Do we still need to vendor `zenoh_c` above?

ahcorde replied:
I think we need to compile `zenoh_cpp_vendor`.

clalancette replied:
Right, my understanding is that the way this works is that `zenoh-cpp` needs either `zenoh-c` or `zenoh-pico` as a "backend" to compile against.  So we need to keep both `zenoh_c_vendor` and `zenoh_cpp_vendor`.

We could consider making this two separate vendor packages, and we might want to do that in the future to enable the easier use of `zenoh-pico` as the backend.  But I think this is OK as-is for now.

Yadunund replied:
> We could consider making this two separate vendor packages, and we might want to do that in the future to enable the easier use of zenoh-pico as the backend. But I think this is OK as-is for now.

One challenge with that would be keeping the commit hashes between the zenoh_c_vendor and zenoh_cpp_vendor compatible. 

Should we just rename this package to `zenoh_vendor` and export c and cpp targets?

clalancette replied:
I think we should leave this as `zenoh_cpp_vendor` for now; after all, that is the API we are working against.  In the future, I think we may want to end up having 3 separate packages: `zenoh_c_vendor`, `zenoh_cpp_vendor`, and `zenoh_pico_vendor`.  But it isn't clear to me how the user would choose between the C and the pico backends, so that would need further thought.

Yadunund commented on rmw_zenoh_cpp/src/zenohd/main.cpp:
```diff
@@ -60,40 +58,36 @@ int main(int argc, char ** argv)
-    "Started Zenoh router with id %s.\n",
-    rmw_zenoh_cpp::liveliness::zid_to_str(z_info_zid(z_session_loan(&session))).c_str());
+
+  std::cout << "Started Zenoh router with id"
+            << session.get_zid().to_string()
+            << std::endl;
+
```
Could we instead use the`RMW_ZENOH_LOG_INFO` logging macro?

ahcorde replied:
This is not ROS code, we just launching the router here

Yadunund commented on rmw_zenoh_cpp/src/zenohd/main.cpp:
```diff
@@ -60,40 +58,36 @@ int main(int argc, char ** argv)
+    zenoh::Session::SessionOptions::create_default(),
+    &result);
+  if (result != Z_OK) {
+    std::cout << "Error opening Session!" << "\\n";
```
Could we instead use the `RMW_ZENOH_LOG_ERROR` logging macro?

ahcorde replied:
Some comment https://github.com/ros2/rmw_zenoh/pull/327#discussion_r1879761955

Yadunund commented on rmw_zenoh_cpp/src/detail/attachment_helpers.hpp:
```diff
@@ -15,34 +15,37 @@
 
 namespace rmw_zenoh_cpp
 {
-///=============================================================================
```
Nit: Undo deletion.

Yadunund commented on rmw_zenoh_cpp/src/detail/attachment_helpers.hpp:
```diff
@@ -15,34 +15,37 @@
+  explicit AttachmentData(
+    const int64_t _sequence_number,
+    const int64_t _source_timestamp,
+    const std::vector<uint8_t> _source_gid);
```
To keep the style consistent, let's not have any leading underscores for variable names.

Yadunund commented on rmw_zenoh_cpp/src/detail/attachment_helpers.hpp:
```diff
@@ -15,34 +15,37 @@
-    const int64_t source_timestamp,
-    const uint8_t source_gid[RMW_GID_STORAGE_SIZE]);
-  explicit AttachmentData(const z_loaned_bytes_t *);
+  explicit AttachmentData(
```
We don't need to define this constructor as `explicit` as there are multiple arguments. 
```suggestion
 AttachmentData(
```

Yadunund commented on rmw_zenoh_cpp/src/detail/attachment_helpers.cpp:
```diff
@@ -27,75 +29,24 @@
-  memcpy(source_gid_, source_gid, RMW_GID_STORAGE_SIZE);
+  sequence_number_ = _sequence_number;
+  source_timestamp_ = _source_timestamp;
+  source_gid_ = _source_gid;
```
Move to initializer list.

Yadunund commented on rmw_zenoh_cpp/src/detail/attachment_helpers.hpp:
```diff
@@ -15,34 +15,37 @@
+    const int64_t _source_timestamp,
+    const std::vector<uint8_t> _source_gid);
+
+  explicit AttachmentData(const zenoh::Bytes & bytes);
```
So that `bytes` can be moved into the Deserializer.
```suggestion
  explicit AttachmentData(zenoh::Bytes bytes);
```

ahcorde replied:
This is already done in the method
```cpp
  zenoh::ext::Deserializer deserializer(std::move(attachment));
````

clalancette replied:
In point of fact, I think that this should be a constref (i.e. it is fine as-is), and further, we shouldn't have the `std::move` inside the method.  That's because inside of the method, what we are doing is taking the the set of bytes, and deserializing.  That's going to involve a copy anyway, so we don't want another copy when the call is made, and further, this method doesn't want to take ownership of the bytes.  So I think this is good as-is, and I'll add another comment for removing that `std::move`.

Yadunund commented on rmw_zenoh_cpp/src/detail/graph_cache.hpp:
```diff
@@ -110,7 +110,7 @@ class GraphCache final
   ///   This is used to infer which entities originated from the current session
   ///   so that appropriate event callbacks may be triggered.
-  explicit GraphCache(const z_id_t & zid);
+  explicit GraphCache(const std::string & zid);
```
Why was this API changed? There seems to be an `Id` object in C++ with a `::to_string()` method we can call in the constructor implementation.
https://zenoh-cpp.readthedocs.io/en/1.0.1/commons.html#_CPPv4N5zenoh2IdE

Yadunund commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -751,7 +751,7 @@ rmw_publisher_assert_liveliness(const rmw_publisher_t * publisher)
   auto pub_data = node_data->get_pub_data(publisher);
   RMW_CHECK_ARGUMENT_FOR_NULL(pub_data, RMW_RET_INVALID_ARGUMENT);
 
-  if (!pub_data->liveliness_is_valid()) {
```
Undo change.

ahcorde commented:
This last [commit](https://github.com/ros2/rmw_zenoh/pull/327/commits/f8e85b1f52d7c0427b01d2f9fd5c8a2b55ce2cdb) from https://github.com/ros2/rmw_zenoh/pull/333 makes `test_services__rmw_zenoh_cpp` flaky

clalancette commented:
> This last [commit](https://github.com/ros2/rmw_zenoh/pull/327/commits/f8e85b1f52d7c0427b01d2f9fd5c8a2b55ce2cdb) from #333 makes `test_services__rmw_zenoh_cpp` flaky

I think we actually need a different fix for that.  I'm going to finish reviewing #333 today, and then I'll open a follow-up PR which should fix that particular problem.

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_client_data.hpp:
```diff
@@ -64,7 +65,7 @@ class ClientData final : public std::enable_shared_from_this<ClientData>
 
   // Copy the GID of this ClientData into an rmw_gid_t.
-  void copy_gid(uint8_t out_gid[RMW_GID_STORAGE_SIZE]) const;
+  std::vector<uint8_t> copy_gid() const;
```
I think we should keep the original function so that we don't have to change any code in rmw_zenoh.cpp

clalancette replied:
My opinion here is that we should actually go with this new signature; while I don't love that the copy is somewhat implicit, that is just C++ rules.  I'm mostly fine with this because the method has the word "copy" in its name.

clalancette replied:
That said, we are noticing additional latency with this port, so we should check to see if this is the cause of it.

clalancette replied:
I think with #340 merged, this is resolved, so resolving this comment.

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.hpp:
```diff
@@ -69,6 +70,7 @@ class PublisherData final
 
   // Copy the GID of this PublisherData into an rmw_gid_t.
   void copy_gid(uint8_t out_gid[RMW_GID_STORAGE_SIZE]) const;
+  std::vector<uint8_t> copy_gid() const;
```
```suggestion
  
  // Return a copy of the GID of this publisher. 
  std::vector<uint8_t> copy_gid() const;
```

Yadunund commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -2451,7 +2451,7 @@ rmw_get_gid_for_publisher(const rmw_publisher_t * publisher, rmw_gid_t * gid)
 
   gid->implementation_identifier = rmw_zenoh_cpp::rmw_zenoh_identifier;
-  pub_data->copy_gid(gid->data);
+  memcpy(gid->data, pub_data->copy_gid().data(), RMW_GID_STORAGE_SIZE);
```
keep original implementation.

clalancette replied:
I think this change is OK, because `copy_gid` is now returning a copy.

Yadunund replied:
I think the change is unnecessary since `PublisherData` has a `copy_gid` overload that fills in the array.

clalancette replied:
Oh, that's a good point.

I'm actually not in favor of having both overloads for `copy_gid`; I think that is just confusing.  I'll add a comment in the `rmw_publisher_data.cpp` class.

Yadunund replied:
I'd prefer keeping the original implementation such that we don't change anything in this file. The `memcpy` becomes in implementation detail.

Yadunund commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -2468,7 +2468,7 @@ rmw_get_gid_for_client(const rmw_client_t * client, rmw_gid_t * gid)
 
   gid->implementation_identifier = rmw_zenoh_cpp::rmw_zenoh_identifier;
-  client_data->copy_gid(gid->data);
+  memcpy(gid->data, client_data->copy_gid().data(), RMW_GID_STORAGE_SIZE);
```
keep original implementation.

clalancette replied:
I think this change is OK, because copy_gid is now returning a copy.

Yadunund commented on rmw_zenoh_cpp/src/rmw_get_topic_endpoint_info.cpp:
```diff
@@ -35,6 +35,7 @@ rmw_get_publishers_info_by_topic(
   bool no_mangle,
   rmw_topic_endpoint_info_array_t * publishers_info)
 {
+  printf("zenoh::rmw_get_publishers_info_by_topic\n");
```
```suggestion
```

Yadunund commented on rmw_zenoh_cpp/src/detail/zenoh_utils.hpp:
```diff
@@ -54,16 +53,16 @@ class ZenohReply final
 
 private:
-  z_owned_query_t query_;
+  std::optional<zenoh::Query> query_;
```
Why is this an optional when the constructor takes in a `zenoh::Query`?

Yadunund commented on rmw_zenoh_cpp/src/detail/zenoh_utils.cpp:
```diff
@@ -52,37 +54,27 @@ std::chrono::nanoseconds::rep ZenohQuery::get_received_timestamp() const
 
 ///=============================================================================
-const z_loaned_query_t * ZenohQuery::get_query() const {return z_loan(query_);}
+const zenoh::Query & ZenohQuery::get_query() const {return query_.value();}
```
Wouldn't this segfault if `query_` is `std::nullopt`?

Yadunund commented on rmw_zenoh_cpp/src/detail/zenoh_utils.hpp:
```diff
@@ -15,37 +15,36 @@
 {
 public:
-  ZenohReply(const z_loaned_reply_t * reply, std::chrono::nanoseconds::rep received_timestamp);
+  ZenohReply(const zenoh::Reply & reply, std::chrono::nanoseconds::rep received_timestamp);
```
Instead of cloning the reply, why don't we std:move a reply into a `ZenohReply` object?

clalancette replied:
Yes, I think that would be more efficient if that works.

clalancette replied:
Actually I think this is fine now.  As a constref, this is basically passing the reply in as a pointer, and then in the implementation of the `ZenohReply` constructor we are doing the clone.  So I think this is only one copy, which we need to do either way.

Yadunund commented on rmw_zenoh_cpp/src/detail/zenoh_router_check.hpp:
```diff
@@ -1,30 +0,0 @@
-// Copyright 2023 Open Source Robotics Foundation, Inc.
```
imo this PR shouldn't make such big changes in this PR. Are we not able to simply port this function to zenoh-cpp?

ahcorde replied:
This is not used anymore, there is something on zenoh_cpp that replace this

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:
```diff
@@ -40,61 +40,23 @@
-//==============================================================================
-// TODO(Yadunund): Make this a class member and lambda capture weak_from_this()
-// instead of passing a rawptr to SubscriptionData when we switch to zenoh-cpp.
-void sub_data_handler(z_loaned_sample_t * sample, void * data)
```
I notice that you've implemented the TODO below but I believe we wanted to do this in a follow up PR after we port the code to zenoh-cpp first. So for now, can we keep these functions and rely on `std::bind` or lambda to call these functions?

ahcorde replied:
I implemented the capture lambdas, implemening the lambdas the code is more easy to follow

clalancette replied:
I think this one is OK to leave as well; we haven't had too much trouble with this one.

clalancette replied:
We decided to stick with the lambdas, so resolving this comment.

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_service_data.hpp:
```diff
@@ -110,12 +111,15 @@ class ServiceData final
+  std::string keyexpr_;
   // An owned queryable.
-  z_owned_queryable_t qable_;
+  std::optional<zenoh::Queryable<void>> qable_;
```
imo this should not be an optional. if we are unable to construct a `zenoh:::Queryable`, the `make()` function should fail.

ahcorde replied:
There is no public contructor for this class

clalancette replied:
This is the same discussion as the earlier one.  Because of the design of these classes, I don't think we can do much about this.  But a comment would be helpful:

```suggestion
  // The Queryable *must* exist in order for anything in this ServiceData class, and hence rmw_zenoh_cpp, to work.
  // However, zenoh::Queryable does not have an empty constructor, so just declaring this as a zenoh::Queryable fails to compile.
  // We work around that by wrapping it in a std::optional, so the std::optional gets constructed at ServiceData constructor time,
  // and then we initialize qable_ later.  Note that the zenoh-cpp API declare_queryable() throws an
  // exception if it fails, so this should all be safe to do.
  std::optional<zenoh::Queryable<void>> qable_;
```

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_service_data.hpp:
```diff
@@ -110,12 +111,15 @@ class ServiceData final
+  std::optional<zenoh::Queryable<void>> qable_;
   // Liveliness token for the service.
-  z_owned_liveliness_token_t token_;
+  std::optional<zenoh::LivelinessToken> token_;
```
Same as above.

ahcorde replied:
There is no public contructor for this class

clalancette replied:
And the same discussion again:

```suggestion
  // The token_ *must* exist in order for anything in this ServiceData class, and hence rmw_zenoh_cpp, to work.
  // However, zenoh::LivelinessToken does not have an empty constructor, so just declaring this as a zenoh::LivelinessToken fails to compile.
  // We work around that by wrapping it in a std::optional, so the std::optional gets constructed at ServiceData constructor time,
  // and then we initialize token_ later.  Note that the zenoh-cpp API liveliness_declare_token() throws an
  // exception if it fails, so this should all be safe to do.
  std::optional<zenoh::LivelinessToken> token_;
```

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -396,14 +418,14 @@ class rmw_context_impl_s::Data final
   std::shared_ptr<rmw_zenoh_cpp::GraphCache> graph_cache_;
   // ROS graph liveliness subscriber.
-  z_owned_subscriber_t graph_subscriber_;
+  std::optional<zenoh::Subscriber<void>> graph_subscriber_cpp_;
```
This should not be an optional. `make()` should fail if we are unable to construct the sub.

ahcorde replied:
There is no public contructor for this class

clalancette replied:
I share @Yadunund 's uncomfortable-ness with this type.  It doesn't seem correct.

However, this won't actually compile if this is just a `zenoh::Subscriber<void>`, because then the constructor for the `Data` class will try to call the empty constructor on `Subscriber`, which doesn't exist.  By wrapping this in an `std::optional`, we workaround this by basically calling the empty constructor for `std::optional` during `Data` construction, and then filling it in for real later on.

The only other way to do this that is coming to mind is to make a new version of `liveliness_declare_subscriber` which returns a pointer, and then to make this a `std::shared_ptr` or something.  But that isn't intrinsically better in my mind.

To fix this for real we'd probably have to restructure the class hierarchy here so that we had one class that contained the `Session`, and another that contained the `Subscriber`.  But I think that is too much to do in this PR, so the only thing I'll request is a comment here explaining why we are doing things this way:

```suggestion
  // The graph_subscriber *must* exist in order for anything in this Data class, and hence rmw_zenoh_cpp, to work.
  // However, zenoh::Subscriber does not have an empty constructor, so just declaring this as a zenoh::Subscriber fails to compile.
  // We work around that by wrapping it in a std::optional, so the std::optional gets constructed at Data constructor time,
  // and then we initialize graph_subscriber_cpp_ later.  Note that the zenoh-cpp API liveliness_declare_subscriber() throws an
  // exception if it fails, so this should all be safe to do.
  std::optional<zenoh::Subscriber<void>> graph_subscriber_cpp_;
```

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -230,22 +240,34 @@ class rmw_context_impl_s::Data final
-      z_undeclare_subscriber(z_move(graph_subscriber_));
-      if (shm_provider_.has_value()) {
-        z_drop(z_move(shm_provider_.value()));
+      // TODO(ahcorde): review this
```
Address todo?

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -417,37 +439,6 @@ class rmw_context_impl_s::Data final
 };
 
-///=============================================================================
-static void graph_sub_data_handler(z_loaned_sample_t * sample, void * data)
```
For this initial port, it would be best if we retained this function while updating the API. We can rely on a lambda or std::bind specify this function as the callback.

clalancette replied:
In this particular case, I'm fine with doing this as a lambda; the code in here is simple enough.  We will eventually need to go through and remove the `data_to_data_shared_ptr` structure in favor of lambda captures, but that will be a separate PR.

clalancette replied:
We decided to stick with the lambdas here, so resolving this.

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -40,72 +43,11 @@
-{
-
-///=============================================================================
-void client_data_handler(z_loaned_reply_t * reply, void * data)
```
Keep these functions for the initial port. We will switch to lambda captures in the follow up PR.

clalancette replied:
Yes, agreed.  This is tricky to get right, so I'd rather do it in a follow-up.

ahcorde replied:
Done here https://github.com/ros2/rmw_zenoh/pull/327/commits/d0da4a6defe563cdf184abbde68479a8938e7819

ahcorde replied:
When I restored this I'm getting some failures in rclcpp tests. I know that this might more complex to review with the lambdas but I believe this is solving some other issues too. I would prefer to keep the other implementation

clalancette replied:
We decided to go with the lambdas, so resolving this comment.

Yadunund reviewed (changes requested):
Left some comments to address from this first pass.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -232,45 +174,27 @@ ClientData::ClientData(
-      NULL
-    ) != Z_OK)
-  {
+  zenoh::ZResult err;
```
Nit: This variable name should reflect a result rather than an error.

YuanYuYuan replied:
Nit: Inappropriate variable name. It's a result rather than an error.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -308,15 +231,14 @@ void ClientData::add_new_reply(std::unique_ptr<ZenohReply> reply)
-      static_cast<int>(z_string_len(z_loan(keystr))),
-      z_string_data(z_loan(keystr)));
+      keystr.size(),
+      keystr.c_str());
```
In zenoh-cpp, the `string_view` can provide the string data with the corresponding length. It's no longer need to create another temp string.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -67,48 +66,57 @@ class rmw_context_impl_s::Data final
             "Unable to connect to a Zenoh router. "
             "Have you started a router with `ros2 run rmw_zenoh_cpp rmw_zenohd`?");
         }
+        zenoh::ZResult err;
```
Nit: Inappropriate variable name as above.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -137,87 +145,89 @@ class rmw_context_impl_s::Data final
+    options.target = zenoh::QueryTarget::Z_QUERY_TARGET_ALL;
+    options.payload = "";
+
+    zenoh::ZResult err;
```
Nit: Inappropriate variable name as above.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_node_data.cpp:
```diff
@@ -58,21 +58,18 @@ std::shared_ptr<NodeData> NodeData::make(
-  z_view_keyexpr_from_str(&liveliness_ke, liveliness_keyexpr.c_str());
-  z_owned_liveliness_token_t token;
-  if (z_liveliness_declare_token(session, &token, z_loan(liveliness_ke), NULL) != Z_OK) {
+  zenoh::ZResult err;
```
Nit: Inappropriate variable name as above.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_node_data.cpp:
```diff
@@ -210,7 +207,7 @@ bool NodeData::create_sub_data(
 
   auto sub_data = SubscriptionData::make(
-    std::move(session),
+    session,
```
Why do we move the `std::shared_ptr<GraphCache>` but not the `std::shared_ptr<zenoh::Session> session`?

YuanYuYuan replied:
Why do we move the `std::shared_ptr<GraphCache>` but not the `std::shared_ptr<zenoh::Session> session`?

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_node_data.cpp:
```diff
@@ -402,7 +399,14 @@ rmw_ret_t NodeData::shutdown()
 
   // Unregister this node from the ROS graph.
-  z_liveliness_undeclare_token(z_move(token_));
+  zenoh::ZResult err;
```
Nit: Inappropriate variable name as above.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -303,24 +273,21 @@ rmw_ret_t PublisherData::publish(
-  z_result_t res = z_publisher_put(z_loan(pub_), z_move(payload), &options);
-  if (res != Z_OK) {
-    if (res == Z_ESESSION_CLOSED) {
+  zenoh::ZResult err;
```
Nit: Inappropriate variable name as above.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_service_data.cpp:
```diff
@@ -316,66 +279,91 @@ rmw_ret_t ServiceData::take_request(
-  // Fill in the request header.
-  // Get the sequence_number out of the attachment
-  AttachmentData attachment(z_query_attachment(loaned_query));
+  auto slice = payload.value().get().slice_iter().next();
```
In zenoh-cpp, we can directly convert it to a vector.

YuanYuYuan replied:
In zenoh-cpp, we can directly convert it to a vector.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_service_data.cpp:
```diff
@@ -447,21 +435,33 @@ rmw_ret_t ServiceData::send_response(
+  std::vector<uint8_t> raw_bytes(
+    reinterpret_cast<const uint8_t *>(response_bytes),
+    reinterpret_cast<const uint8_t *>(response_bytes) + data_length);
+  zenoh::Bytes payload(raw_bytes);
```
Actually we should use `zenoh::Bytes payload(std::move(raw_bytes));` to avoid a copy. See [here](https://github.com/eclipse-zenoh/zenoh-cpp/blob/94edfd9845b63bab4761be328bd64db95daab212/include/zenoh/api/bytes.hxx?plain=1#L65) for details.

YuanYuYuan replied:
Actually we should use `zenoh::Bytes payload(std::move(raw_bytes));` to avoid a copy. See [here](https://github.com/eclipse-zenoh/zenoh-cpp/blob/94edfd9845b63bab4761be328bd64db95daab212/include/zenoh/api/bytes.hxx?plain=1#L65) for details.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -346,47 +268,67 @@ rmw_ret_t ClientData::take_response(
-    reinterpret_cast<char *>(const_cast<uint8_t *>(z_slice_data(z_loan(payload)))),
-    z_slice_len(z_loan(payload)));
+  auto & payload = sample.get_payload();
+  auto slice = payload.slice_iter().next();
```
Use `as_vector`.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -308,15 +231,14 @@ void ClientData::add_new_reply(std::unique_ptr<ZenohReply> reply)
     // Log warning if message is discarded due to hitting the queue depth
-    z_view_string_t keystr;
-    z_keyexpr_as_view_string(z_loan(keyexpr_), &keystr);
+    std::string keystr = std::string(keyexpr_.value().as_string_view());
```
In zenoh-cpp, the `string_view` can provide the string data with the corresponding length. It's no longer need to create another temp string.

ahcorde replied:
I'm getting this error 

```bash
conversion from ‘std::string_view’ {aka ‘std::basic_string_view<char>’} to non-scalar type ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} requested
  236 |     std::string keystr = keyexpr_.value().as_string_view();
```

YuanYuYuan replied:
Answered in https://github.com/ros2/rmw_zenoh/pull/327#issuecomment-2536435747

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -458,23 +394,64 @@ rmw_ret_t ClientData::send_request(
+  std::vector<uint8_t> raw_bytes(
+    reinterpret_cast<const uint8_t *>(request_bytes),
+    reinterpret_cast<const uint8_t *>(request_bytes) + data_length);
+  opts.payload = zenoh::Bytes(raw_bytes);
```
We should use `zenoh::Bytes(std::move(raw_bytes))` to avoid a copy. See https://github.com/eclipse-zenoh/zenoh-cpp/blob/94edfd9845b63bab4761be328bd64db95daab212/include/zenoh/api/bytes.hxx?plain=1#L65 for details.

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:
```diff
@@ -473,37 +495,48 @@ rmw_ret_t SubscriptionData::take_one_message(
 
-  const uint8_t * payload = z_slice_data(z_loan(msg_data->payload));
-  const size_t payload_len = z_slice_len(z_loan(msg_data->payload));
+  auto slice = msg_data->payload.slice_iter().next();
```
Use `as_vector`.

ahcorde committed:
Merge remote-tracking branch 'origin/rolling' into ahcorde/dev/1.0.0-cpp

ahcorde committed:
Added feedback

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

ahcorde commented:
| Pacakge                 | Test                                                                  | Reason                                                    |
| -                       | -                                                                     | -                                                         |
| rcl | None                             |                                            |
| rcl_action | None                             |                                            |
| rcl_lifecycle | None                             |                                            |
| rcl_yaml | None                             |                                            |
| rcl_yaml_param_parser | None                             |                                            |
| rclcpp | test_executors                             |                                            |
| rclcpp_components | None                             |                                            |
| rclcpp_action | None                             |                                            |
| rclcpp_lifecycle | None                             |                                            |
| rclcpp_components | None                             |                                            |

ahcorde commented:
Fixed test_quality_of_service tests https://github.com/ros2/system_tests/pull/555

YuanYuYuan commented:
> test_executors

This matches what we observed so far. I'm still investigating the flaky test_executors. And the test_n_nodes__rmw_zenoh_cpp has been identified https://github.com/ros2/rmw_zenoh/issues/329

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/zenoh_utils.cpp:
```diff
@@ -39,23 +41,23 @@ ZenohSession::~ZenohSession()
-  AttachmentData data(sequence_number, source_timestamp, gid);
-  data.serialize_to_zbytes(out_bytes);
+  rmw_zenoh_cpp::AttachmentData data(sequence_number, source_timestamp, gid);
+  return std::move(data.serialize_to_zbytes());
```
The move is superfluous and causes the following warning.

```console
/workspace/src/rmw_zenoh/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp: In function ‘zenoh::Bytes rmw_zenoh_cpp::create_map_and_set_sequence_num(int64_t, std::vector<unsigned char>)’:
/workspace/src/rmw_zenoh/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp:52:19: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
   52 |   return std::move(data.serialize_to_zbytes());
      |          ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/src/rmw_zenoh/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp:52:19: note: remove ‘std::move’ call
---
Finished <<< rmw_zenoh_cpp [8.39s]

```

Covered in https://github.com/ros2/rmw_zenoh/pull/327#issuecomment-2536435747.

YuanYuYuan commented:
@ahcorde I made a suggested change here.

```diff
diff --git a/rmw_zenoh_cpp/src/detail/rmw_client_data.cpp b/rmw_zenoh_cpp/src/detail/rmw_client_data.cpp
index bae0fe9..3cad22c 100644
--- a/rmw_zenoh_cpp/src/detail/rmw_client_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_client_data.cpp
@@ -233,14 +233,13 @@ void ClientData::add_new_reply(std::unique_ptr<ZenohReply> reply)
     reply_queue_.size() >= adapted_qos_profile.depth)
   {
     // Log warning if message is discarded due to hitting the queue depth
-    std::string keystr = std::string(keyexpr_.value().as_string_view());
+    // std::string keystr = std::string(keyexpr_.value().as_string_view());
     RMW_ZENOH_LOG_ERROR_NAMED(
       "rmw_zenoh_cpp",
       "Query queue depth of %ld reached, discarding oldest Query "
-      "for client for %.*s",
+      "for client for %s",
       adapted_qos_profile.depth,
-      keystr.size(),
-      keystr.c_str());
+      keyexpr_.value().as_string_view());
     reply_queue_.pop_front();
   }
   reply_queue_.emplace_back(std::move(reply));
@@ -407,10 +406,11 @@ rmw_ret_t ClientData::send_request(
     parameters,
     [client_data](const zenoh::Reply & reply) {
       if (!reply.is_ok()) {
+        auto reply_err_str = reply.get_err().get_payload().as_string();
         RMW_ZENOH_LOG_ERROR_NAMED(
           "rmw_zenoh_cpp",
           "z_reply_is_ok returned False Reason: %s",
-          reply.get_err().get_payload().as_string())
+          reply_err_str.c_str())
         return;
       }
       const zenoh::Sample & sample = reply.get_ok();
@@ -420,7 +420,7 @@ rmw_ret_t ClientData::send_request(
         RMW_ZENOH_LOG_ERROR_NAMED(
           "rmw_zenoh_cpp",
           "Unable to obtain ClientData from data for %s.",
-          std::string(sample.get_keyexpr().as_string_view()));
+          sample.get_keyexpr().as_string_view());
         return;
       }
 
diff --git a/rmw_zenoh_cpp/src/detail/rmw_service_data.cpp b/rmw_zenoh_cpp/src/detail/rmw_service_data.cpp
index 5696e81..d24c965 100644
--- a/rmw_zenoh_cpp/src/detail/rmw_service_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_service_data.cpp
@@ -156,7 +156,7 @@ std::shared_ptr<ServiceData> ServiceData::make(
         RMW_ZENOH_LOG_ERROR_NAMED(
           "rmw_zenoh_cpp",
           "Unable to obtain ServiceData from data for %s.",
-          std::string(query.get_keyexpr().as_string_view()));
+          query.get_keyexpr().as_string_view());
         return;
       }
 
diff --git a/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp b/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
index 45ef1bf..a6d0114 100644
--- a/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
@@ -212,7 +212,7 @@ bool SubscriptionData::init()
           RMW_ZENOH_LOG_ERROR_NAMED(
             "rmw_zenoh_cpp",
             "Unable to obtain SubscriptionData from data for %s.",
-            std::string(sample.get_keyexpr().as_string_view()));
+            sample.get_keyexpr().as_string_view());
           return;
         }
 
diff --git a/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp b/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
index 795c4dc..3e0f6a4 100644
--- a/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
+++ b/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
@@ -49,7 +49,7 @@ zenoh::Bytes create_map_and_set_sequence_num(
   int64_t source_timestamp = now_ns.count();
 
   rmw_zenoh_cpp::AttachmentData data(sequence_number, source_timestamp, gid);
-  return std::move(data.serialize_to_zbytes());
+  return data.serialize_to_zbytes();
 }
 
 ///=============================================================================

```

ahcorde committed:
Added feedback

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

ahcorde commented:
Thank you  @YuanYuYuan, included here https://github.com/ros2/rmw_zenoh/pull/327/commits/6877a48f6ae4e088eda4f832e62d6d73746b70b0

YuanYuYuan commented on rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:
```diff
@@ -201,14 +163,15 @@ SubscriptionData::SubscriptionData(
-  std::string topic_keyexpr = entity_->topic_info()->topic_keyexpr_;
-  z_view_keyexpr_t sub_ke;
-  if (z_view_keyexpr_from_str(&sub_ke, topic_keyexpr.c_str()) != Z_OK) {
+  if (entity_->topic_info()->qos_.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE) {
+    RMW_ZENOH_LOG_WARN_NAMED(
+      "rmw_zenoh_cpp",
+      "`reliability` no longer supported on subscriber. Ignoring...");
+  }
```
It has been removed from the rolling since the 1.0 porting. We've decided that this information is annoying and only useful for the developers.

ahcorde replied:
https://github.com/ros2/rmw_zenoh/pull/327/commits/8a419851a3b0ef96bfd54d69f51f30f2b1d0b4c7

ahcorde committed:
Removed annoying message

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

ahcorde committed:
Merge remote-tracking branch 'origin/rolling' into ahcorde/dev/1.0.0-cpp

clalancette commented on rmw_zenoh_cpp/src/detail/attachment_helpers.cpp:
```diff
@@ -122,17 +73,38 @@ size_t AttachmentData::gid_hash() const
-  ze_serializer_serialize_str(z_loan_mut(serializer), "source_gid");
-  ze_serializer_serialize_buf(z_loan_mut(serializer), this->source_gid_, RMW_GID_STORAGE_SIZE);
-  ze_serializer_finish(z_move(serializer), attachment);
+  zenoh::ext::Deserializer deserializer(std::move(bytes));
```
```suggestion
  zenoh::ext::Deserializer deserializer(bytes);
```

(we don't want to take ownership of these bytes)

clalancette commented on rmw_zenoh_cpp/src/detail/liveliness_utils.cpp:
```diff
@@ -679,7 +678,18 @@ std::string demangle_name(const std::string & input)
 
 ///=============================================================================
-size_t hash_gid(const uint8_t gid[RMW_GID_STORAGE_SIZE])
+size_t hash_gid(const std::vector<uint8_t> gid)
```
I'm not a big fan of having both `hash_gid` and `hash_gid_p`.  Is there any way we can just do one of these?  Barring that, can we make one call the other?

clalancette replied:
This comment is still open.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -244,46 +176,27 @@ ClientData::ClientData(
 
 ///=============================================================================
-bool ClientData::init(std::shared_ptr<ZenohSession> session)
+bool ClientData::init(const std::shared_ptr<zenoh::Session> session)
```
This is an existing problem, but we may as well fix it; since we are now passing the session into the `ClientData` constructor, we don't need to pass it here.  We already have the session in `sess_`.

```suggestion
bool ClientData::init()
```

clalancette replied:
(further, I think that given what we have changed elsewhere we might be able to get rid of this `init` method completely, but I'm going to wait until we do the lambda captures before asking to remove this)

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -244,46 +176,27 @@ ClientData::ClientData(
-    ) != Z_OK)
-  {
+  zenoh::ZResult result;
+  this->token_ = session->liveliness_declare_token(
```
```suggestion
  this->token_ = sess_->liveliness_declare_token(
```

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -244,46 +176,27 @@ ClientData::ClientData(
-      z_drop(z_move(this->token_));
-    });
 
   sess_ = session;
```
```suggestion
```

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.hpp:
```diff
@@ -107,14 +108,14 @@ class ClientData final : public std::enable_shared_from_this<ClientData>
 
   // Initialize the Zenoh objects for this entity.
-  bool init(std::shared_ptr<ZenohSession> session);
+  bool init(const std::shared_ptr<zenoh::Session> session);
```
```suggestion
  bool init();
```

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -244,46 +176,27 @@ ClientData::ClientData(
-
+  std::string topic_keyexpr = this->entity_->topic_info().value().topic_keyexpr_;
+  keyexpr_ = zenoh::KeyExpr(topic_keyexpr);
+  // TODO(ahcorde) check KeyExpr
```
Do we still need to do something here?

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -145,90 +148,89 @@ class rmw_context_impl_s::Data final
+    graph_subscriber_cpp_ = session_->liveliness_declare_subscriber(
+      keyexpr_cpp,
+      [&](const zenoh::Sample & sample) {
+        // Look up the data shared_ptr in the global map. If it is in there, use it.
+        // If not, it is being shutdown so we can just ignore this update.
+        std::shared_ptr<rmw_context_impl_s::Data> data_shared_ptr{nullptr};
+        {
+          std::lock_guard<std::mutex> lk(data_to_data_shared_ptr_map_mutex);
+          if (data_to_data_shared_ptr_map.count(this) == 0) {
+            return;
+          }
+          data_shared_ptr = data_to_data_shared_ptr_map[this];
+        }
```
This is one of those things that is going to be able to go away with lambda captures available.  But I think we should leave this as-is for now, and do a separate PR fixing it.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -263,29 +227,32 @@ rmw_ret_t PublisherData::publish(
-    RMW_CHECK_FOR_NULL_WITH_MSG(
-      msg_bytes, "bytes for message is null", return RMW_RET_BAD_ALLOC);
-  }
+  // if (shm_provider.has_value()) {
+  //   RMW_ZENOH_LOG_DEBUG_NAMED("rmw_zenoh_cpp", "SHM is enabled.");
+
+  //   auto provider = shm_provider.value()._0;
+  //   z_buf_layout_alloc_result_t alloc;
+  //   // TODO(yuyuan): SHM, configure this
+  //   z_alloc_alignment_t alignment = {5};
+  //   z_shm_provider_alloc_gc_defrag_blocking(
+  //     &alloc,
+  //     z_loan(provider),
+  //     SHM_BUF_OK_SIZE,
+  //     alignment);
+
+  //   if (alloc.status == ZC_BUF_LAYOUT_ALLOC_STATUS_OK) {
+  //     shmbuf = std::make_optional(alloc.buf);
+  //     msg_bytes = reinterpret_cast<char *>(z_shm_mut_data_mut(z_loan_mut(alloc.buf)));
+  //   } else {
+  //     // TODO(Yadunund): Should we revert to regular allocation and not return an error?
+  //     RMW_SET_ERROR_MSG("Failed to allocate a SHM buffer, even after GCing.");
+  //     return RMW_RET_ERROR;
+  //   }
+  // } else {
```
What's the deal with this?  Does zenoh-cpp not support SHM yet?

ahcorde replied:
To remove the C API I commented this part, I added a comment to rewirte this with zenoh_cpp API

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -355,19 +319,18 @@ rmw_ret_t PublisherData::publish_serialized_message(
+  auto options = zenoh::Publisher::PutOptions::create_default();
+  options.attachment = create_map_and_set_sequence_num(sequence_number_++, entity_->copy_gid());
+
+  std::vector<uint8_t> raw_image(
+    serialized_message->buffer,
+    serialized_message->buffer + data_length);
+  zenoh::Bytes payload(raw_image);
```
I have a minor and a major question to ask here.

The minor thing is that I don't love the term "raw_image" here, as this is any data, not just images.  Maybe "raw_data" instead?

The more major thing I want to ask about this is how many copies we are doing with this sequence.  Does constructing a `std::vector` with a pointer and length make a copy, or just refer to the pointer?  And then does passing a `std::vector` into the `zenoh::Bytes` constructor take a reference, or make a copy?

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_service_data.cpp:
```diff
@@ -450,21 +432,33 @@ rmw_ret_t ServiceData::send_response(
+    request_id->sequence_number,
+    writer_gid);
+
+  std::vector<uint8_t> raw_bytes(
+    reinterpret_cast<const uint8_t *>(response_bytes),
+    reinterpret_cast<const uint8_t *>(response_bytes) + data_length);
+  zenoh::Bytes payload(std::move(raw_bytes));
```
Same question as above; how many copies of the data is this?

clalancette reviewed (changes requested):
I've left quite a few changes inline here, but this is looking pretty good.

ahcorde committed:
Feedback

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

ahcorde committed:
Removed shm and small fix

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

ahcorde commented:
@YuanYuYuan related to the string_view 

I need to use in `rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:239`

```cpp
      std::string(keyexpr_.value().as_string_view()).c_str());
```

Otherwise I got a segfault in `rclcpp/test_client_common`. Any idea why ?

ahcorde committed:
Merge remote-tracking branch 'origin/rolling' into ahcorde/dev/1.0.0-cpp

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -359,47 +266,63 @@ rmw_ret_t ClientData::take_response(
-    RMW_SET_ERROR_MSG("Failed to get source_timestamp from client call attachment");
+  // Object that manages the raw buffer
+  auto payload = sample.get_payload().as_vector();
+  if (payload.size() > 0) {
```
We can save an indentation level here by doing:

```
if (payload.size() == 0) {
    RMW_ZENOH_LOG_DEBUG_NAMED(
      "rmw_zenoh_cpp",
      "ClientData not able to get slice data");
    return RMW_RET_ERROR;
}
```

(and then unindenting the rest of the code)

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -359,47 +266,63 @@ rmw_ret_t ClientData::take_response(
+      return RMW_RET_ERROR;
+    }
+
+    // Fill in the request_header
+    if (!sample.get_attachment().has_value()) {
+      RMW_ZENOH_LOG_DEBUG_NAMED(
+        "rmw_zenoh_cpp",
+        "ClientData take_request attachment is empty");
+      return RMW_RET_ERROR;
+    }
```
This is a really minor point, but in the case that a message doesn't have an attachment, we wasted time above doing the deserialization.  Thus my suggestion is that we move this check right after the "payload.size == 0" check above, and get out early.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -359,47 +266,63 @@ rmw_ret_t ClientData::take_response(
+      RMW_SET_ERROR_MSG("Failed to get source_timestamp from client call attachment");
+      return RMW_RET_ERROR;
+    }
+    memcpy(
+      request_header->request_id.writer_guid,
+      attachment.copy_gid().data(),
+      RMW_GID_STORAGE_SIZE);
```
I'm just noticing that this is doing two copies; once when `attachment.copy_gid()` is called, and then again during the memcpy.  I think we should reduce this back to one copy.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -471,23 +388,66 @@ rmw_ret_t ClientData::send_request(
+      }
+
+      std::chrono::nanoseconds::rep received_timestamp =
+      std::chrono::system_clock::now().time_since_epoch().count();
```
```suggestion
        std::chrono::system_clock::now().time_since_epoch().count();
```

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -471,23 +388,66 @@ rmw_ret_t ClientData::send_request(
+      sub_data->add_new_reply(
+        std::make_unique<rmw_zenoh_cpp::ZenohReply>(reply, received_timestamp));
+    },
+    [client_data]() {
+      auto sub_data = client_data.lock();
+      if (sub_data == nullptr) {
+        RMW_ZENOH_LOG_ERROR_NAMED(
+          "rmw_zenoh_cpp",
+          "Unable to obtain ClientData");
+        return;
+      }
+    },
```
Hm, I'm kind of thinking that we don't need a `data_drop` callback at all anymore.  Given the fact that we are now capturing the weak pointer for the data callback, we know that either the ClientData object is alive, or not.  We no longer need to do any additional tracking.  In other words, I think this can now be a "null" callback.

clalancette commented on rmw_zenoh_cpp/src/detail/zenoh_utils.hpp:
```diff
@@ -41,26 +42,24 @@ class ZenohSession final
 
 private:
-  z_owned_reply_t reply_;
+  std::optional<zenoh::Reply> reply_;
```
Does this need to be an `optional`?  I kind of think that we always have to have one of these (given the constructor), but it may be the case that we can't construct an empty `zenoh::Reply` object.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -401,6 +333,12 @@ void PublisherData::copy_gid(uint8_t out_gid[RMW_GID_STORAGE_SIZE]) const
   entity_->copy_gid(out_gid);
 }
 
+std::vector<uint8_t> PublisherData::copy_gid() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  return entity_->copy_gid();
+}
```
I'm not a fan of having two overloads for `copy_gid` here.  Can we reduce this to one?

clalancette reviewed (changes requested):
All right, I did another pass on this.  This is looking pretty close to me, with a few more relatively minor things to fix.  The only remaining major issue is to see if we can reduce the latency with this vs. the `rolling` branch.

ahcorde committed:
Added feedback

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

YuanYuYuan commented:
> `test_client_common`

I see. I observed similar issues before. Maybe the macro doesn't work well with the lifetime of the keyexpr. In this case, we need to store the string additionally.

YuanYuYuan committed:
Merge pull request #340 from ZettaScaleLabs/fix/sub-high-latency

Fix the high latency on the subscriber due to zenoh bytes conversion

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:
```diff
@@ -240,19 +196,51 @@ bool SubscriptionData::init()
+
+        sub_data->add_new_message(
+          std::make_unique<SubscriptionData::Message>(
+            sample.get_payload().as_vector(),
```
Is this safe to do?  The sample is a `constref` here, and we are basically taking a pointer.  Will that data still exist when we go to dereference it during `take_one_message` later?  How are we guaranteed that?

YuanYuYuan replied:
The function name `as_vector` may not be appropriate... `as_vector` copies and stores the data in a `std::vector` and then we move it into the `Message`. I've tested it with a ping pong test. No misbehavior is found.

clalancette replied:
I see, OK.  I guess what `as_vector` does is to copy the data from the internal list of bytes to a vector, and return it.  So the vector will outlive the Sample.  In that case, this is fine.

ahcorde commented:
When compiling on Windows I'm getting this error just include the zenoh.hxx header

<details>
<summary><b>Windows compile error</b></summary>
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(36,1): error C2236: token inesperado 'struct'. Compruebe si olvid� un ';' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(36,1): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(36,24): error C2513: '': no hay ninguna variable declarada antes de '=' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(37,1): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(37,25): error C2226: error de sintaxis: tipo 'zenoh::ShmProviderAsyncInterface::drop::<unnamed-tag>' inesperado [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(41,1): error C2236: token inesperado 'struct'. Compruebe si olvid� un ';' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(41,1): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(41,24): error C2513: '': no hay ninguna variable declarada antes de '=' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(42,1): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\protocol_implementations\posix\../../provider/shm_provider.hxx(42,18): error C2059: error de sintaxis: '->' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(38,10): error C2236: token inesperado 'struct'. Compruebe si olvid� un ';' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(38,20): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(38,20): error C2513: '': no hay ninguna variable declarada antes de '=' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(39,14): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(39,14): error C2059: error de sintaxis: '->' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(43,10): error C2236: token inesperado 'struct'. Compruebe si olvid� un ';' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(43,20): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(43,20): error C2513: '': no hay ninguna variable declarada antes de '=' [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(44,21): error C2332: 'struct': falta el nombre de etiqueta [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
C:\ros2_ws\install\opt\zenoh_cpp_vendor\include\zenoh\api\shm\provider\alloc_layout.hxx(44,21): error C2226: error de sintaxis: tipo 'zenoh::shm::provider::closures::_z_alloc_layout_async_interface_drop_fn::<unnamed-tag>' inesperado [C:\ros2_ws\build\rmw_zenoh_cpp\rmw_zenohd.vcxproj]
Failed   <<< rmw_zenoh_cpp [4.02s, exited with code 1]

</details>

clalancette committed:
Switch to storing GIDs with a std::array.

They are a fixed size, so should use the fixed-size data
structure.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Remove unnecessary ClientData init() method.

Since we aren't calling shared_from_this() in it anymore,
we can do all of the work in the constructor.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Small refactorings.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

ahcorde committed:
Fixed windows

Signed-off-by: Alejandro Hernandez Cordero <ahcorde@gmail.com>

clalancette committed:
Fix bad string print.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Rewrite connecting to the router to only print when it fails.

This simplifies the code, and makes it less noisy.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette reviewed (approved):
This now looks good to me, so I'm going to go ahead and merge it.  Thanks for all of the hard work here @ahcorde

reinzor commented:
Just a FYI, there are still quite some `zenoh-c` references. In the readme, package.xml etc.
